server/meca/article.js

```diff
@@ -0,0 +1,185 @@
+    formatter: editorFormatter,
+  },
+  'potential-reviewers': {
+    keys: ['suggestedReviewer', 'opposedReviewer'],
+    formatter: member => ({
+      name: {
+        // TODO split name?
+        surname: member.meta.name,
+        'given-names': '',
+      },
+      email: member.meta.email,
+    }),
+  },
+}
+
+const articleTypeMap = {
+  'research-article': 5,
+  'short-report': 13,
+  'tools-resources': 19,
+  'scientific-correspondence': 20,
+  feature: 23,
+}
+
+function createXmlObject(manuscript, editorsById, affiliations) {
+  return {
+    'journal-meta': {
+      'journal-id': [
+        { '@journal-id-type': 'pmc' },
+        { '@journal-id-type': 'publisher' },
+      ],
+      issn: {},
+      publisher: {
+        'publisher-name': {},
+      },
+    },
+
+    'article-meta': {
+      'article-id': {
+        '@pub-id-type': 'manuscript',
+        '#text': manuscript.id,
+      },
+      'article-categories': {
+        'subj-group': [].concat(
+          {
+            '@subj-group-type': 'article-type',
+            subject: manuscript.meta.articleType,
+          },
+          {
+            '@subj-group-type': 'article-type-id',
+            subject: articleTypeMap[manuscript.meta.articleType],
+          },
+          manuscript.meta.subjects.map(subject => ({
+            'subj-group': { '@subj-group-type': 'keywords', subject },
+          })),
+        ),
+      },
+      'title-group': {
+        'article-title': manuscript.meta.title,
+      },
+
+      'contrib-group': Object.entries(contribStructure).map(
+        ([group, groupInfo]) => ({
+          '@content-type': group,
+          contrib: groupInfo.keys.reduce(
+            (contribs, role) =>
+              contribs.concat(
+                manuscript.teams
+                  .find(team => team.role === role)
+                  .teamMembers.map(member => ({
+                    '@contrib-type': lodash.kebabCase(role),
+                    ...groupInfo.formatter(member, affiliations, editorsById),
+                  })),
+              ),
+            [],
+          ),
+        }),
+      ),
+
+      aff: affiliations.map((aff, index) => ({
+        '@id': `aff${index}`,
+        institution: aff,
+      })),
+
+      'author-notes': { fn: { p: {} } },
+      'pub-date': { day: '', month: '', year: '' },
+
+      'related-article': manuscript.cosubmission.map(title => ({
+        '@related-article-type': 'companion',
+        'article-title': title,
+      })),
+
+      'custom-meta-group': {
+        'custom-meta': Object.entries({
+          'suggestions-conflict': manuscript.suggestionsConflict,
+          'previously-discussed': manuscript.previouslyDiscussed,
+          'previously-submitted': manuscript.previouslySubmitted[0],
+          'created-by': manuscript.createdBy,
+        })
+          .filter(([name, value]) => value)
+          .map(([name, value]) => ({ 'meta-name': name, 'meta-value': value })),
+      },
+    },
+  }
+}
```

Comment:
This is quite confusing to read. It may be worth refactoring out some of the nested logic elsewhere and spreading. 

Have we considered creating a model in the long term that we can use serverside, and share with the meca component which can provide an XML view for it?

Reply:
Agreed. I'll have a go at splitting it up.

This is specifically a MECA XML view of the manuscript though so I think it makes more sense in the MECA component than in the Manuscript entity manager.

Reply:
I split the file in two. It's not much better though. I'm going to merge it now but feel free to have a go at refactoring it again if you can see a way to improve it.